tests/unit/test_charm.py

```diff
@@ -439,150 +440,101 @@ def test_unit_ips(self):
+    ):
+        # preset values
+        self.harness.set_leader(True)
+        single_replica.return_value.is_ready.return_value = True
+
+        # test cases where replica set should and shouldn't be reconfigured
+        for reconfiguration in [False, True]:
+            self.harness.charm._need_replica_set_reconfiguration = reconfiguration
+
+            # simulate 2nd MongoDB unit
+            rel = self.harness.charm.model.get_relation("mongodb")
+            rel_id = rel.id
+            key_values = {"private-address": "127.4.5.6"}
+            self.harness.add_relation_unit(rel_id, "mongodb/1")
+            self.harness.update_relation_data(rel_id, "mongodb/1", key_values)
+
+            # check if mongod reconfigured replica set
+            if reconfiguration:
+                mongodb_reconfigure.assert_called()
+
+                # verify replica set hosts updated accordingly
+                self.assertEqual(
+                    self.harness.charm._replica_set_hosts,
+                    ["127.4.5.6", "1.1.1.1"],
+                )
 
-            # verify that we do not initialise replica set
-            initialise_replica_set.assert_not_called()
+            else:
+                mongodb_reconfigure.assert_not_called()
 
     @patch_network_get(private_address="1.1.1.1")
-    @patch("mongod_helpers.MongoDB.is_replica_set")
-    @patch("charm.MongodbOperatorCharm._initialise_replica_set")
-    @patch("mongod_helpers.MongoDB.is_ready")
-    @patch("charm.MongodbOperatorCharm.app")
-    def test_mongodb_relation_joined_peers_ready(
-        self, app, is_ready, initialise_replica_set, is_replica_set
+    @patch("charm.MongodbOperatorCharm._single_mongo_replica")
+    @patch("charm.MongodbOperatorCharm._need_replica_set_reconfiguration")
+    @patch("mongod_helpers.MongoDB.reconfigure_replica_set")
+    def test_mongodb_relation_joined_failure(
+        self, mongodb_reconfigure, need_reconfiguration, single_replica
     ):
         # preset values
         self.harness.set_leader(True)
-        is_ready.return_value = True
-        is_replica_set.return_value = False
+        single_replica.return_value.is_ready.return_value = True
+        self.harness.charm._need_replica_set_reconfiguration = True
 
-        # simulate 2nd MongoDB unit
-        rel = self.harness.charm.model.get_relation("mongodb")
-        rel_id = rel.id
-        key_values = {"private-address": "127.4.5.6"}
-        self.harness.add_relation_unit(rel_id, "mongodb/1")
-        self.harness.update_relation_data(rel_id, "mongodb/1", key_values)
-        app.planned_units.return_value = len(self.harness.charm._unit_ips)
+        # test cases where replica set should and shouldn't be reconfigured
+        exceptions = [
+            ConnectionFailure("error message"),
+            ConfigurationError("error message"),
+            OperationFailure("error message"),
+        ]
+        for exception in exceptions:
+            # exception
```

Comment:
This comment doesn't really add any extra information

Reply:
LOL oh god 🤦